File: sdks/tokenizer-sdk/src/accounts/asset.ts
/**
 * Asset account deserializer.
 *
 * #[repr(C)] layout (368 bytes):
 *   0: account_key (u8)
 *   1: version (u8)
 *   4: id (u32)                    — padded from 2
 *   8: organization ([u8;32])
 *  40: collection ([u8;32])
 *  72: total_shares (u64)
 *  80: minted_shares (u64)
 *  88: status (u8)
 *  89: transfer_policy (u8)
 *  96: price_per_share (u64)       — padded from 90
 * 104: accepted_mint ([u8;32])
 * 136: dividend_epoch (u32)
 * 140: fundraising_round_count (u32)
 * 144: created_at (i64)
 * 152: updated_at (i64)
 * 160: bump (u8)
 * 161: collection_authority_bump (u8)
 * 162: native_treasury ([u8;32])
 * 194: active_buyout ([u8;32])
 * 228: unminted_succeeded_rounds (u32) — padded from 226
 * 232: open_distributions (u32)
 * 236: compliance_program ([u8;32])
 * 272: transfer_cooldown (i64)     — padded from 268
 * 280: max_holders (u32)
 * 284: current_holders (u32)
 * 288: maturity_date (i64)
 * 296: maturity_grace_period (i64)
 * 304: oracle_source (u8)
 * 305: _oracle_pad ([u8;7])
 * 312: oracle_feed ([u8;32])
 * 344: shares_per_unit (u64)
 * 352: last_oracle_update (i64)
 * 360: oracle_max_staleness (u32)
 * 364: oracle_max_confidence_bps (u16)
 * 366: accepted_mint_decimals (u8)
 * 367: _oracle_reserved (u8)
 */
import { type Address, getStructDecoder, transformDecoder } from "gill";
import { AccountKey, type AssetStatus, type OracleSource } from "../constants.js";
import { addr, i64d, pad, u8d, u16d, u32d, u64d } from "./decode.js";

export const ASSET_SIZE = 368;
export const ASSET_OFFSET_ORGANIZATION = 8;

export interface Asset {
  accountKey: number;
  version: number;
  id: number;
  organization: Address;
  collection: Address;
  totalShares: bigint;
  mintedShares: bigint;
  status: AssetStatus;
  transferPolicy: number;
  pricePerShare: bigint;
  acceptedMint: Address;
  dividendEpoch: number;
  fundraisingRoundCount: number;
  createdAt: bigint;
  updatedAt: bigint;
  bump: number;
  collectionAuthorityBump: number;
  nativeTreasury: Address;
  activeBuyout: Address;
  unmintedSucceededRounds: number;
  openDistributions: number;
  complianceProgram: Address;
  transferCooldown: bigint;
  maxHolders: number;
  currentHolders: number;
  maturityDate: bigint;
  maturityGracePeriod: bigint;
  oracleSource: OracleSource;
  oracleFeed: Address;
  sharesPerUnit: bigint;
  lastOracleUpdate: bigint;
  oracleMaxStaleness: number;
  oracleMaxConfidenceBps: number;
  acceptedMintDecimals: number;
}

const rawDecoder = getStructDecoder([
  ["accountKey", u8d],
  ["version", u8d],
  ["_p0", pad(2)],
  ["id", u32d],
  ["organization", addr],
  ["collection", addr],
  ["totalShares", u64d],
  ["mintedShares", u64d],
  ["status", u8d],
  ["transferPolicy", u8d],
  ["_p1", pad(6)],
  ["pricePerShare", u64d],
  ["acceptedMint", addr],
  ["dividendEpoch", u32d],
  ["fundraisingRoundCount", u32d],
  ["createdAt", i64d],
  ["updatedAt", i64d],
  ["bump", u8d],
  ["collectionAuthorityBump", u8d],
  ["nativeTreasury", addr],
  ["activeBuyout", addr],
  ["_p2", pad(2)],
  ["unmintedSucceededRounds", u32d],
  ["openDistributions", u32d],
  ["complianceProgram", addr],
  ["_p3", pad(4)],
  ["transferCooldown", i64d],
  ["maxHolders", u32d],
  ["currentHolders", u32d],
  ["maturityDate", i64d],
  ["maturityGracePeriod", i64d],
  // Oracle fields
  ["oracleSource", u8d],
  ["_p4", pad(7)],
  ["oracleFeed", addr],
  ["sharesPerUnit", u64d],
  ["lastOracleUpdate", i64d],
  ["oracleMaxStaleness", u32d],
  ["oracleMaxConfidenceBps", u16d],
  ["acceptedMintDecimals", u8d],
  ["_p5", pad(1)],
]);

export const assetDecoder = transformDecoder(
  rawDecoder,
  ({ _p0, _p1, _p2, _p3, _p4, _p5, status, transferPolicy, oracleSource, ...rest }) => ({
    ...rest,
    status: status as AssetStatus,
    transferPolicy,
    oracleSource: oracleSource as OracleSource,
  }),
);

export function decodeAsset(data: Uint8Array): Asset {
  if (data.length < ASSET_SIZE) {
    throw new Error(`Asset: expected ${ASSET_SIZE} bytes, got ${data.length}`);
  }
  const accountKey = data[0];
  if (accountKey !== AccountKey.Asset) {
    throw new Error(`Asset: invalid account key ${accountKey}`);
  }
  return assetDecoder.decode(data);
}

File: programs/tokenizer/src/state/asset.rs
use pinocchio::Address;

use super::{AssetStatus, ASSET_SEED};

/// Tokenizable asset with an associated Metaplex Core collection.
/// Metadata (name, URI) lives in the collection — not duplicated here.
/// Token count (num_minted) is tracked by the collection natively.
/// PDA: ["asset", organization.key(), &id.to_le_bytes()]
#[repr(C)]
pub struct Asset {
    // Discriminator
    pub account_key: u8,              // AccountKey::Asset
    pub version: u8,

    // Identity
    pub id: u32,
    pub organization: [u8; 32],      // Parent organization PDA

    // Metaplex Core
    pub collection: [u8; 32],        // Metaplex Core collection address

    // Shares
    pub total_shares: u64,
    pub minted_shares: u64,

    // State
    pub status: u8,                   // AssetStatus
    pub transfer_policy: u8,          // TransferPolicy (0 = NonTransferable, 1 = Transferable)

    // Pricing
    pub price_per_share: u64,         // Smallest stablecoin unit
    pub accepted_mint: [u8; 32],     // Which stablecoin

    // Dividends
    pub dividend_epoch: u32,          // Current dividend epoch

    // Fundraising
    pub fundraising_round_count: u32, // Number of rounds created for this asset

    // Timestamps
    pub created_at: i64,
    pub updated_at: i64,

    // PDA
    pub bump: u8,

    // Collection authority PDA bump (cached for CPI signing)
    pub collection_authority_bump: u8,

    /// Native treasury wallet for this asset's governance (set by create_asset_governance).
    /// Zero = no governance, funds go to org authority.
    pub native_treasury: [u8; 32],

    // Terms & Conditions fields

    /// Active buyout PDA (future buyout use). Zero = no active buyout.
    pub active_buyout: [u8; 32],

    /// Number of succeeded rounds whose tokens haven't been fully minted yet.
    pub unminted_succeeded_rounds: u32,

    /// Number of open (unclosed) dividend distributions.
    pub open_distributions: u32,

    /// Future Arcium compliance program address. Zero = none.
    pub compliance_program: [u8; 32],

    /// Minimum seconds between transfers for any token of this asset. 0 = no cooldown.
    pub transfer_cooldown: i64,

    /// Maximum number of distinct holders. 0 = unlimited.
    pub max_holders: u32,

    /// Current number of distinct holders.
    pub current_holders: u32,

    /// Unix timestamp after which asset is matured (no new rounds). 0 = no maturity.
    pub maturity_date: i64,

    /// Grace period (seconds) after maturity_date before asset expires. 0 = no grace.
    pub maturity_grace_period: i64,

    // Oracle pricing (opt-in)

    /// Oracle source: 0 = None (manual pricing), 1 = Pyth, 2 = Switchboard
    pub oracle_source: u8,
    /// Reserved for alignment after oracle_source
    pub _oracle_pad: [u8; 7],
    /// Oracle price feed account pubkey. Zeroed if oracle_source == 0.
    pub oracle_feed: [u8; 32],
    /// How many shares represent 1 unit of the underlying asset.
    /// e.g., 1000 means 1000 shares = 1 oz gold.
    /// Oracle price is divided by this to get price_per_share.
    pub shares_per_unit: u64,
    /// Unix timestamp of the last oracle price refresh.
    pub last_oracle_update: i64,
    /// Maximum staleness in slots before oracle price is rejected.
    pub oracle_max_staleness: u32,
    /// Maximum confidence/std_dev in basis points (e.g., 200 = 2%).
    /// 0 = skip confidence check.
    pub oracle_max_confidence_bps: u16,
    /// Mint decimals for accepted_mint (cached to avoid extra account read on refresh).
    pub accepted_mint_decimals: u8,
    /// Reserved for future oracle config.
    pub _oracle_reserved: u8,
}

impl Asset {
    pub const LEN: usize = core::mem::size_of::<Self>();

    #[inline(always)]
    pub unsafe fn load(data: &[u8]) -> &Self {
        &*(data.as_ptr() as *const Self)
    }

    #[inline(always)]
    pub unsafe fn load_mut(data: &mut [u8]) -> &mut Self {
        &mut *(data.as_mut_ptr() as *mut Self)
    }

    pub fn derive_pda(
        organization: &[u8; 32],
        id: u32,
        program_id: &Address,
    ) -> (Address, u8) {
        Address::find_program_address(
            &[ASSET_SEED, organization, &id.to_le_bytes()],
            program_id,
        )
    }

    #[inline(always)]
    pub fn status(&self) -> AssetStatus {
        // Safe because we validate on write
        unsafe { core::mem::transmute(self.status) }
    }

    /// Returns true if the asset has reached its maturity date.
    #[inline(always)]
    pub fn is_matured(&self, now: i64) -> bool {
        self.maturity_date != 0 && now >= self.maturity_date
    }

    /// Returns true if the asset has passed maturity + grace period.
    #[inline(always)]
    pub fn is_expired(&self, now: i64) -> bool {
        self.maturity_date != 0
            && self.maturity_grace_period != 0
            && now >= self.maturity_date.saturating_add(self.maturity_grace_period)
    }
}
